File: updated_code/code/java/1351-count-negative-numbers-in-a-sorted-matrix.java
// Package declaration
package com.example;

import java.util.Arrays;

/**
 * This class solves the problem of counting negative numbers in a sorted matrix.
 */
public class Solution {
    /**
     * This method counts the number of negative numbers in a sorted matrix.
     * 
     * @param grid the input matrix
     * @return the number of negative numbers in the matrix
     */
    public int countNegatives(int[][] grid) {
        if (grid == null || grid.length == 0 || grid[0].length == 0) {
            return 0;
        }
        
        int row = grid.length;
        int col = grid[0].length;
        int count = 0;
        
        // Start from the top right corner and move diagonally down and left
        int i = 0, j = col - 1;
        while (i < row && j >= 0) {
            if (grid[i][j] < 0) {
                // If the current element is negative, move down and increment the count
                count += col - j;
                i++;
            } else {
                // If the current element is not negative, move left
                j--;
            }
        }
        
        return count;
    }
}

File: updated_code/code/python/1351-count-negative-numbers-in-a-sorted-matrix.py
import typing

class Solution:
    def count_negatives(self, grid: typing.List[typing.List[int]]) -> int:
        """
        Given a `m x n` matrix `grid` which is sorted in non-increasing order both
        row-wise and column-wise, return the number of **negative** numbers in `grid`.

        Example 1:
            Input: grid = [[4, 3, 2, -1], [3, 2, 1, -1], [1, 1, -1, -2], [-1, -1, -2, -3]]
            Output: 8
            Explanation: There are 8 negatives number in the matrix.

        Example 2:
            Input: grid = [[3, 2], [1, 0]]
            Output: 0

        Constraints:
            * `m == grid.length`
            * `n == grid[i].length`
            * `1 <= m, n <= 100`
            * `-100 <= grid[i][j] <= 100`

        Follow up: Could you find an `O(n + m)` solution?
        """
        m, n = len(grid), len(grid[0])
        i, j = 0, n-1
        count = 0
        while i < m and j >= 0:
            if grid[i][j] < 0:
                count += n - j
                i += 1
            else:
                j -= 1
        return count

# Final newline added